common/archiver/s3store/historyArchiver.go

```diff
@@ -223,38 +267,61 @@ func (h *historyArchiver) Get(
 		}
 		token = &getHistoryToken{
 			CloseFailoverVersion: *highestVersion,
-			NextBatchIdx:         0,
 		}
 	}
 
-	key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion)
-	encodedHistoryBatches, err := download(ctx, h.s3cli, URI, key)
+	prefix := constructHistoryKeyPrefixWithVersion(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion)
+
+	results, err := h.s3cli.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket:     aws.String(URI.Hostname()),
+		Prefix:     aws.String(prefix),
+		MaxKeys:    aws.Int64(int64(request.PageSize)), // Upper bound of history events
+		StartAfter: (*token).StartAfterKey,
+	})
+
 	if err != nil {
-		return nil, err
+		if isRetryableError(err) {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		return nil, &shared.BadRequestError{Message: err.Error()}
 	}
-	historyBatches, err := decodeHistoryBatches(encodedHistoryBatches)
-	if err != nil {
-		return nil, &shared.InternalServiceError{Message: err.Error()}
+	if len(results.Contents) == 0 {
+		return nil, &shared.BadRequestError{Message: errNoHistoryBatchesFound}
```

Comment:
Please return `&shared.EntityNotExistError{Message: XXX}` in this case.

Reply:
See next comment. I'm not sure if it's the intended behaviour to retry. This specific error seems pretty hard to trigger though